solver/src/main/java/org/chocosolver/solver/constraints/nary/PropDiffN.java

```diff
@@ -136,51 +162,78 @@ protected void filterFromBox(int i) throws ContradictionException {
-                filter(i, j, horizontal);
+
+        if (xLengthMin > 0 && yLengthMin > 0) {
+            int maxNumberRectangles = ((xM - xm) / xLengthMin) * ((yM - ym) / yLengthMin);
+            if (maxNumberRectangles < overlappingBoxes.getNeighOf(i).size()+1) {
+                fails();
             }
-            assert !(doOverlap(i, j, horizontal) && doOverlap(i, j, vertical));
         }
     }
 
+    private boolean mayOverlap(int i, int j) {
+        return isNotDisjoint(i, j, true) && isNotDisjoint(i, j, false);
+    }
+
+    private boolean isNotDisjoint(int i, int j, boolean horizontal) {
+        int off = (horizontal) ? 0 : n;
+        return (vars[i + off].getLB() < vars[j + off].getUB() + vars[j + off + 2 * n].getUB())
+            && (vars[j + off].getLB() < vars[i + off].getUB() + vars[i + off + 2 * n].getUB());
+    }
+
     private boolean doOverlap(int i, int j, boolean hori) {
         int offSet = hori ? 0 : n;
-        int S_i = vars[i + offSet].getUB();
+        int s_i = vars[i + offSet].getUB();
         int e_i = vars[i + offSet].getLB() + vars[i + 2 * n + offSet].getLB();
-        int S_j = vars[j + offSet].getUB();
+        int s_j = vars[j + offSet].getUB();
         int e_j = vars[j + offSet].getLB() + vars[j + 2 * n + offSet].getLB();
-        return (S_i < e_i && e_j > S_i && S_j < e_i)
-                || (S_j < e_j && e_i > S_j && S_i < e_j);
+        return (s_i < e_i && e_j > s_i && s_j < e_i)
+            || (s_j < e_j && e_i > s_j && s_i < e_j);
     }
 
-    private void filter(int i, int j, boolean hori) throws ContradictionException {
+    private boolean filter(int i, int j, boolean hori) throws ContradictionException {
+        boolean hasFiltered = false;
         int offSet = hori ? 0 : n;
-        int S_i = vars[i + offSet].getUB();
+        int s_i = vars[i + offSet].getUB();
         int e_i = vars[i + offSet].getLB() + vars[i + 2 * n + offSet].getLB();
-        int S_j = vars[j + offSet].getUB();
+        int s_j = vars[j + offSet].getUB();
         int e_j = vars[j + offSet].getLB() + vars[j + 2 * n + offSet].getLB();
-        if (S_i < e_i || S_j < e_j) {
-            if (e_j > S_i) {
-                vars[j + offSet].updateLowerBound(e_i, this);
-                vars[i + offSet].updateUpperBound(S_j - vars[i + 2 * n + offSet].getLB(), this);
-                vars[i + offSet + 2 * n].updateUpperBound(S_j - vars[i + offSet].getLB(), this);
+        if (s_i < e_i || s_j < e_j) {
+            if (e_j > s_i) {
+                if(vars[j + offSet].updateLowerBound(e_i, this)) {
+                    if(!pruneList.contains(j)) {
+                        pruneList.add(j);
+                    }
+                    hasFiltered = true;
+                }
+                if(vars[i + offSet].updateUpperBound(s_j - vars[i + 2 * n + offSet].getLB(), this)
+                    || vars[i + offSet + 2 * n].updateUpperBound(s_j - vars[i + offSet].getLB(), this)) {
```

Comment:
beware, if first bound update returns true, the second one will not be applied, resulting in less filtering